spec.bs

```diff
@@ -201,68 +208,250 @@ This document introduces a new storage API that is intentionally not partitioned
+    1. If the result of running [=determine whether shared storage is allowed by enrollment and user preference=] given |outsideSettings| and [=this=]'s [=SharedStorageWorklet/worklet origin=] is false:
+        1. If [=this=]'s [=SharedStorageWorklet/is cross-origin worklet=] is false, [=reject=] |promise| with a {{TypeError}}.
+        1. Else, [=resolve=] |promise| with undefined.
+        1. Return |promise|.
+    1. Return |promise|, and immediately [=obtaining a worklet agent=] given |outsideSettings| and run the rest of these steps in that agent:
+        1. [=Queue a global task=] on the [=DOM manipulation task source=], given |window|, to [=resolve=] |promise| with undefined.
+        1. If [=this=]'s [=module map=] is not [=map/empty=]:
+            1. Let |operationMap| be [=this=]'s {{SharedStorageWorkletGlobalScope}}'s [=SharedStorageWorkletGlobalScope/operation map=].
+            1. If |operationMap| [=map/contains=] |name|:
+                1.  Let |operation| be |operationMap|[|name|].
+                1. If |options| [=map/contains=] |data|:
+                    1. Let |argumentsList| be a new [=/list=].
+                    1. [=list/Append=] |data| to |argumentsList|.
+                    1. [=Invoke=] |operation| with |argumentsList|.
+                1. Otherwise, [=invoke=] |operation| without any arguments list.
+        1. If |options|["`keepAlive`"] is false:
+            1. Wait for |operation| to finish running, if applicable.
+            1. Run [=terminate a worklet global scope=] with [=this=]'s {{SharedStorageWorkletGlobalScope}}.
+  </div>
+
   ## Monkey Patch for [=Worklets=] ## {#worklet-monkey-patch}
 
   This specification will make some modifications to the [=Worklet=] standard to accommodate the needs of Shared Storage.
 
-  In particular, the {{Worklet/addModule()}} method steps for {{Worklet}} will need to be prepended with the following step:
+  In particular, the {{Worklet/addModule()}} method steps for {{Worklet}} will need to include the following step before the step "Let |promise| be a new promise":
+
+    4. If |this| is of type {{SharedStorageWorklet}}:
+        1. Let |addModuleAllowedResult| be the result of running [=check if addModule is allowed and update state=] given |this| and <var ignore=''>moduleURLRecord</var>.
+        1. If |addModuleAllowedResult| is "DisallowedDueToNonPreferenceError":
+            1. Reject <var ignore="">promise</var> with a {{TypeError}}.
+        1. Else if |addModuleAllowedResult| is "DisallowedDueToPreferenceError":
+            1. If |this|'s [=SharedStorageWorklet/is cross-origin worklet=] is false, then reject <var ignore="">promise</var> with a {{TypeError}}. Else, resolve <var ignore="">promise</var>.
+        1. Else:
+            1. [=Assert=]: |addModuleAllowedResult| is "Allowed".
+            1. Resolve <var ignore="">promise</var>.
+        1. Return <var ignore="">promise</var>.
```

Comment:
If I have asked this before, them I'm sorry. Do we really always unconditionally return a promise right away in the `SharedStorageWorklet` case? It looks like we'd never run the rest of the steps in https://html.spec.whatwg.org/#dom-worklet-addmodule, and therefore we'd never fetch the module script or create the global (and store it to the list of global scopes) or anything. Am I missing something?

This is kind of related to a concern I have about calling the "obtain a worklet agent" algorithm manually in both of the select and run helper algorithms. Calling that algorithm, I think, will create a fresh worklet agent from scratch each time. This agent/realm/execution context trio will be completely different from the agent associated with the SINGLE global scope that exists in the "list of global scopes" (for SharedStorageWorkletGlobalScope). Furthermore, we're storing functions from some(??) agent in the operations map, and then calling them in this new freshly-created agent, which is bad. Let me know if that doesn't make sense, or if I'm off-base somewhere.

Reply:
> and the global scope should be associated with a single [operation map](https://wicg.github.io/shared-storage/#sharedstorageworkletglobalscope-operation-map). And thus, the references in the spec seem legitimate in general.

This is very useful to know (and I probably knew it but forgot, sadly). Can you add an assertion to every place where we pull an IDL callback function out of the operations map and invoke it, to assert that the function object's [associated realm](https://dontcallmedom.github.io/webdex/a.html#associated%20realm%40%40webidl%25%25dfn) (see https://webidl.spec.whatwg.org/#dfn-associated-realm) matches the [relevant realm](https://html.spec.whatwg.org/multipage/webappapis.html#concept-relevant-realm) of the global object?

> can different agents access the same global scope (from the spec's perspective)?

Hmm, I'm curious: why is this question relevant? It sounds like there is only ever a single shared storage worklet agent created for each hosting document, right?

Reply:
Added the assert.

> Hmm, I'm curious: why is this question relevant?

The [obtain a worker/worklet agent](https://html.spec.whatwg.org/multipage/webappapis.html#obtaining-a-worker/worklet-agent) algorithm creates an agent each time. So it seems it each worklet corresponds to a single "agent cluster", but there could be multiple "agents".